topology: Make removenode use left_token_ring state for global barrier

emdroid opened the pull request:
Make the removenode operation go through the `left_token_ring` state, similar to decommission. This ensures that when removenode completes, all nodes in the cluster are aware of the topology change through a global token metadata barrier.

Previously, removenode would skip the `left_token_ring` state and go directly from `write_both_read_new` to `left` state. This meant that when the operation completed, some nodes might not yet know about the topology change, potentially causing issues with subsequent data plane requests.

Key changes:
- Both decommission and removenode now transition to `left_token_ring` state in the `write_both_read_new` handler
- In `left_token_ring` state, only decommissioning nodes receive the shutdown RPC (removed nodes are already dead)
- Updated documentation to reflect that both operations use this state

This change improves consistency guarantees for removenode operations by ensuring cluster-wide awareness before completion.

The change is protected by "REMOVENODE_WITH_LEFT_TOKEN_RING" feature flag to also support mixed clusters during e.g. upgrade.

Fixes:  scylladb/scylladb#25530

No backport: This fixes and issue found in tests. It can theoretically happen in production too, but wasn't reported in any customer issue, so a backport is not needed.

emdroid commented on service/topology_coordinator.cc:
```diff
@@ -2647,29 +2647,20 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
+                    builder.set_transition_state(topology::transition_state::left_token_ring)
+                           .set_version(_topo_sm._topology.version + 1)
                            .with_node(node.id)
-                           .del("tokens")
```
Not sure about this one yet - decommissioned node removing the tokens here ... should the removenode do the same? (but currently having failures when done for removing node as well)

patjed41 commented:
> > Previously, removenode would skip the left_token_ring state and go directly from write_both_read_new to left state. This meant that when the operation completed, some nodes might not yet know about the topology change, potentially causing issues with subsequent data plane requests.
> 
> Bit confused! In case of decommissioning node, IIUC, the global barrier was needed so that other nodes stop contacting the decommission node for requests. But since in case of remove node, the node is already dead/out of contact, what difference does this barrier make?

You are right that we don't care that much about requests sent to a removed node. The node is dead and banned at some point, so requests sent to it will safely fail. The issue here is that some node may still send requests to the removed node after removenode completes. This is an unexpected behavior that causes problems in tests, and even users could encounter it in theory.

If you don't understand how removenode can complete before all nodes stop sending requests to the removed node, read this discussion:  https://github.com/scylladb/scylla-dtest/pull/6325#discussion_r2469271432.

patjed41 reviewed (commented):
It looks like using the cluster feature is not that bad, but waiting in the removenode initiator could still be safer/simpler.

Only several line changes impact removenode with the new feature disabled, but I would still test it (at least manually) just to be sure. All non-upgrade tests will start running with the feature enabled, so a bug here could go undetected.

gusev-p commented on db/system_keyspace.cc:
```diff
@@ -3403,7 +3403,10 @@ static bool must_have_tokens(service::node_state nst) {
+    // REMOVENODE_WITH_LEFT_TOKEN_RING feature is enabled. To support both
+    // cases, we allow removing nodes to not have tokens.
+    case service::node_state::removing: return false;
     case service::node_state::rebuilding: return true;
```
Too many changes in one commit. Would be way simpler to review if the changes were split into smaller commits with clear commit messages. Probably we can take "Key changes:" section and create a separate commit for each bullet.

gusev-p commented on service/topology_coordinator.cc:
```diff
@@ -2806,6 +2826,7 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
                 bool barrier_failed = false;
                 // Wait until other nodes observe the new token ring and stop sending writes to this node.
                 try {
+                    // For removenode this should work as well - the removed node is added to ignored nodes, so it should be automatically excluded.
```
Same here -- we can just `SCYLLA_ASSERT` this invariant

gleb-cloudius commented on docs/dev/topology-over-raft.md:
```diff
@@ -141,7 +142,9 @@ reads that started before this point exist in the system. Finally we remove the
-`commit_cdc_generation`.
+`commit_cdc_generation`. Both decommission and removenode go through the
+`left_token_ring` state to run a global barrier ensuring all nodes are aware
+of the topology change before the operation completes.
```
I do not see that you updated the transition diagram. Did you?

emdroid replied:
good catch, updated the transition diagram

emdroid commented:
New version:
* updated the transition diagram
* asserts for invariants instead of just comments
* fixed commit messages
* after the barrier failure, only sleep for decommission

[diff](https://github.com/scylladb/scylladb/compare/8f6f3f960803ec543fcaab82557a4d7ae29df79a..9a17377e74ab0d2f98402be142a88d936c21b7d1)

emdroid commented on service/storage_service.cc:
```diff
@@ -576,12 +576,14 @@ future<storage_service::nodes_to_notify_after_sync> storage_service::sync_raft_t
+        case node_state::removing:
+            // A decommissioning or removing node loses its tokens when topology moves to left_token_ring.
+            // Note thate we do not need to check for the REMOVENODE_WITH_LEFT_TOKEN_RING feature for removing nodes here,
+            // as seeing `transition_state::left_token_ring` for removenode operation means that the feature is enabled
```
removed

emdroid commented on service/topology_coordinator.cc:
```diff
@@ -2758,6 +2764,11 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
                 auto node = get_node_to_work_on(std::move(guard));
 
+                // Note that we don't need to check the `REMOVENODE_WITH_LEFT_TOKEN_RING` feature here - if it's disabled,
+                // we won't be in `left_token_ring` state in the first place (for the removenode operation).
```
Updated to use the `on_internal_error()` (there is also `on_fatal_internal_error()` that is more equivalent to what `SCYLLA_ASSERT()` does, i.e. abort, but it seems that just throwing an exception is sufficient here). The internal error messages also make the error more clear so less comments are needed (if any at all).

gleb-cloudius replied:
> 
> Why do we even have `SCYLLA_ASSERT` then? Shouldn't we redefine it in terms of `on_internal_error()`? There is too much confusion now to chose between various assert functions.

IIRC current wisdom is to use `SCYLLA_ASSERT` in case not aborting may cause data corruption and use `on_internal_error` otherwise.

emdroid commented on service/topology_coordinator.cc:
```diff
@@ -2823,15 +2844,15 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
 
                 if (barrier_failed) {
-                    // If barrier above failed it means there may be unfinished writes to a decommissioned node.
+                    // If barrier above failed it means there may be unfinished writes to a decommissioned/removed node.
```
fixed, the problem was apparently using some attributes of the node when the node instance might be no longer valid

patjed41 commented on service/topology_coordinator.cc:
```diff
@@ -2781,6 +2787,13 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
                 const auto node_rs_state = node.rs->state;
                 const auto node_id = node.id;
 
+                const bool is_removenode = node_rs_state == node_state::removing;
```
A bit strange to have `is_removenode` when we already have `node_rs_state`. `is_removenode` is used in 4 places, so maybe it's fine.

patjed41 commented on service/topology_coordinator.cc:
```diff
@@ -2828,8 +2853,13 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
+                    if (is_removenode && !excluded_nodes.contains(node_id)) {
+                        on_internal_error(rtlogger, "removenode operation must have the removed node in excluded_nodes");
+                    }
+                    node = retake_node(co_await global_token_metadata_barrier(std::move(node.guard), excluded_nodes), node_id);
```
`std::move(excluded_nodes)`?

emdroid commented on service/topology_coordinator.cc:
```diff
@@ -2777,6 +2777,10 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
 
+                // Need to be captured as the node variable might become invalid (e.g. moved out) at particular points.
+                const auto node_rs_state = node.rs->state;
+                const auto node_id = node.id;
```
Yeah, I first thought the node.id was the issue, but it was actually the node.rs ... so I suppose we don't need to capture the node.id (but it might still be "cleaner" to capture it, as the id shouldn't change during the state processing anyway)

emdroid replied:
reverted the unnecessary `node.id` changes for minimal diff

patjed41 commented on service/topology_coordinator.cc:
```diff
@@ -2842,18 +2846,20 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
-                    node = retake_node(co_await start_operation(), node.id);
+                    if (node_rs_state == node_state::decommissioning) {
+                        // If barrier above failed it means there may be unfinished writes to a decommissioned node.
+                        // Lets wait for the ring delay for those writes to complete and new topology to propagate
```
If we revert this change, we should also update the comment. It shouldn't be only about unfinished writes.

emdroid commented:
rebased

gusev-p reviewed (commented):
Shouldn't we add an explicit cluster/test that removenode waits for all nodes to observe the latest state?

emdroid commented:
> Shouldn't we add an explicit cluster/test that removenode waits for all nodes to observe the latest state?

Well, we are fixing the "update_cluster_layout_tests.TestUpdateClusterLayout.test_simple_removenode_2" test with this change, but it only failed randomly ... we could think about a targeted test.py test , but would probably need to find a way to reproduce this consistently (other than adding delays to the raft library that made the dtest reproduce with higher ratio). In particular, the original code race window is fairly small and I'm currently not sure if we could find a good place to put e.g. an error injection (the mutation to move the node to left state is done before making the request complete, but for consistent repro we'd probably need to have the opposite).

So it's verified locally that the change fixes the test_simple_removenode_2() dtest (using a temporary change as in https://github.com/scylladb/scylladb/issues/25530#issuecomment-3422786650), and I also performed an upgrade test (besides the ones we actually have, as there are existing upgrade tests that do removenode and decommission), but I currently don't have a good idea how to devise a consistent reproducer in the test.py.
